### ISSUE TITLE
updating to deep-cover 0.7.1 will break rspec

### ISSUE THREAD

**Issue by zs787** [deep-cover/deep-cover#48]
Im just wondering why adding deep-cover 0.7.1 will break rspec, resulting in the following error.
```
group :test do
  gem 'deep-cover', require: false
```
```
if ENV['COVERAGE'] || ENV['CI']
  require 'deep_cover/builtin_takeover'
  require 'simplecov'
```
```
An error occurred while loading ./spec/api/api_spec.rb.
Failure/Error: require File.expand_path('../config/environment', __dir__)

NameError:
  method `exec' not defined in PG::Connection
.
.
.
```
we are using ruby 2.5.3 

I installed deep-cover before and the default version that was installed is 0.5 and everything ran with no problems. Then after upgrading to 0.7.1 two of our repos would fail to run rspec resulting in the above problem.

Removed the deep-cover gem from the gem file and everything was running normally again, with no problems.

**Comment by MaxLap** [comment]
Thank you for the report!

This is odd. Could you give us the full content of your Gemfile.lock as well as the full backtrace of the exception? It would help us identify and reproduce the problem.

**Comment by marcandre** [comment]
Very odd indeed.

This happens even if `DeepCover` is not required at all (i.e. neither `ENV['COVERAGE'] || ENV['CI']`)?

The only dependency change from 0.5 to 0.7.1 was that we added `thor`. You could try having `gem 'deep-cover-core', require: false` in your Gemfile instead, just to see if that's the issue somehow.

**Comment by zs787** [comment]
> Very odd indeed.
> 
> This happens even if `DeepCover` is not required at all (i.e. neither `ENV['COVERAGE'] || ENV['CI']`)?
> 
> The only dependency change from 0.5 to 0.7.1 was that we added `thor`. You could try having `gem 'deep-cover-core', require: false` in your Gemfile instead, just to see if that's the issue somehow.

Yes only happens when I require deep-cover, when I take that line out it works fine with no issues.

**Comment by zs787** [comment]
> Thank you for the report!
> 
> This is odd. Could you give us the full content of your Gemfile.lock as well as the full backtrace of the exception? It would help us identify and reproduce the problem.

Thank you for the response, the rest of team here are trying to see if they can figure out anything. If we cant we will try to see what we can share since the files are from our production product.

**Comment by marcandre** [comment]
> Thank you for the response, the rest of team here are trying to see if they can figure out anything. If we cant we will try to see what we can share since the files are from our production product.

Cool.

Note that if it is `DeepCover` rewriting incorrectly the code (as it appears to be), it can be difficult to figure it out. 

Two tricks I could recommend to isolate which code is being badly rewritten:

### restrict instrumentation to one or a few files:

```
require 'deep_cover/builtin_takeover'
DeepCover.configure { paths("#{__dir__}../config/application.rb") }
```
(you can pass an array to `paths`, and paths can be globs)

### or not instrumenting a particular method:

I believe the following works for *not* instrumenting a method `foo` within an instrumented file:

```
# surround
def foo
  ...
end
# with
class_eval <<-'RUBY'
def foo
...
end
RUBY
```

So for example the first thing to try could be to uninstrument just the method where the error is raised. If you get no error (or a different error somewhere else), then the instrumentation of `foo` looks problematic. If you can send us just the code of that method we can probably figure it out from there.

If you get the same error, then the instrumentation error is upstream, so uninstrument the caller, etc...

Let us know if we can help.

**Comment by zs787** [comment]
We have just realized the following, while looked at this issue with another colleague.

Before we used 0.5.2 and it was working as expected.
After updating to 0.7.1 as I mentioned earlier our rspec started to fail and the error above was observed.

inspecting the two directories for 0.7.1 and 0.5.2 we see that the path ```require 'deep_cover/builtin_takeover'``` does not exist under the deep-cover lib for 0.7.1
Downgrading and using 0.5.2 we see no problems and rspec works just fine.
Please note that we are using this with "simplecov", so im using the suggested setup for that as per the readme


```
$ pwd
.../.gem/ruby/2.5.3/gems/deep-cover-0.7.1/lib/deep_cover
$ ls -l
cli
cli.rb
cover_cloned_tree.rb
expression_debugger.rb
instrumented_clone_reporter.rb

$ pwd
.../.gem/ruby/2.5.3/gems/deep-cover-0.5.2/lib/deep_cover 
$ ls -l
analyser
analyser.rb
auto_run.rb
autoload_tracker.rb
backports.rb
base.rb
basics.rb
builtin_takeover.rb
cli
config.rb
config_setter.rb
core_ext
coverage
coverage.rb
covered_code.rb
custom_requirer.rb
flag_comment_associator.rb
load.rb
memoize.rb
module_override.rb
node
node.rb
parser_ext
problem_with_diagnostic.rb
reporter
reporter.rb
tools
tools.rb
version.rb
```

**Comment by MaxLap** [comment]
Hello, this is normal.

The deep-cover gem depends on deep-cover-core. deep-cover-core also set things in the deep-cover name-space. This is not the common way to do things for gems, but it is fully supported.

You can verify this by doing `puts $LOADED_FEATURES.grep(/takeover/)` after doing the `require 'deep_cover/builtin_takeover`.

Wish I could help, but without at least a backtrace it's difficult. You can remove the lines which refer to your application (leaving ... to indicate it), so that only the lines in the gems are included.

**Comment by zs787** [comment]
@MaxLap Thanks again for your response.

Unfortunately beyond the following error happening for all tests I have nothing to provide.
Once the ```require deep_cover...```  line is added to my spec file the following error happens for every single test we have.

```
An error occurred while loading ./.../.../xxx.rb.
Failure/Error: require File.expand_path('../config/environment', __dir__)

NameError:
 method `exec' not defined in PG::Connection
# ./vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:283:in `remove_method'
# ./vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:283:in `block in async_api='
# ./vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:282:in `each'
# ./vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:282:in `async_api='
# ./vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:289:in `<class:Connection>'
# ./vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:21:in `<top (required)>'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `block in require'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:257:in `load_dependency'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
# ./vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg.rb:68:in `<module:PG>'
# ./vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg.rb:35:in `<top (required)>'
# ./config/application.rb:9:in `<top (required)>'
# ./config/environment.rb:4:in `require_relative'
# ./config/environment.rb:4:in `<top (required)>'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `block in require'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:257:in `load_dependency'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
# ./spec/rails_helper.rb:6:in `<top (required)>'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `block in require'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:257:in `load_dependency'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
# ./.../xxx_spec.rb:3:in `<top (required)>'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:285:in `load'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:285:in `block in load'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:257:in `load_dependency'
# ./vendor/bundle/ruby/2.5.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:285:in `load'
/home/.../vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg.rb:38: warning: already initialized constant PG::VERSION
/home/.../vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg.rb:38: warning: previous definition of VERSION was here
/home/.../vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg.rb:41: warning: already initialized constant PG::REVISION
/home/.../vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg.rb:41: warning: previous definition of REVISION was here
/home/.../vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:24: warning: already initialized constant PG::Connection::CONNECT_ARGUMENT_ORDER
/home/.../vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:24: warning: previous definition of CONNECT_ARGUMENT_ORDER was here
/home/.../vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:271: warning: already initialized constant PG::Connection::REDIRECT_METHODS

/home/.../vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/pg/connection.rb:271: warning: previous definition of REDIRECT_METHODS was here
```

Please let me know if there is anything else I can do on my end based on the above.

Thanks

**Comment by marcandre** [comment]
That backtrace helps a lot.

The actual code causing an issue doesn't seem problematic, although we're going to double check.

One thing stands out though: it looks like the `pg` gem is being loaded twice, once from `./vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib/` and once from `/home/.../vendor/bundle/ruby/2.5.0/gems/pg-1.1.4/lib`. It may be related.

I wonder: are you doing `bundle exec rspec ...`? With `bundle exec`, you shouldn't get those warnings.

Of course, that wouldn't explain why you'd get the error with one version of deep-cover vs another, but maybe by upgrading deep-cover it also changed the version of the `pg` gem you're using?

**Comment by zs787** [comment]
> I wonder: are you doing bundle exec rspec ...? With bundle exec, you shouldn't get those warnings.

Yes, I use  ```bundle exec rspec```

> One thing stands out though: it looks like the pg gem is being loaded twice

locally and in our pipeline we have the same error results, other branches are running tests with no problems locally and in the pipeline. I will have a look at this one.

**Comment by marcandre** [comment]
> > One thing stands out though: it looks like the pg gem is being loaded twice
> 
> locally and in our pipeline we have the same error results, other branches are running tests with no problems locally and in the pipeline. I will have a look at this one.

Interesting. Do you have the same warnings even when deep-cover is not being `require`d?

**Comment by zs787** [comment]
> Interesting. Do you have the same warnings even when deep-cover is not being required?

as far as I have seen before no, just when using the deep-cover gem.

**Comment by marcandre** [comment]
The trail gets warmer:
1) I doubled checked, and deep-cover's instrumentation of that method is not the issue
2) this code is executed from `PG.Connection::asynch_api=true`; this is done when `pg` is required (like in the backtrace) or afterwards by the user. The first time it is called, the `remove_method` is not supposed to be called, only when called afterwards by the user. Clearly looks like there's some double loading going on, and that Ruby is somehow confused.
3)A change from deep-cover 0.5 to 0.7 is that we now use the Ruby loading hook `load_iseq`. I suspect this is the change that is involved here with the double loading and/or confusion of Ruby.

I still wonder what the bug is. I'm trying to see if I can reproduce it locally. I'm not super knowledgeable as to how come there are gems in `./vendor/` and in `/home/...` though. Is this run with `bundle exec --deployment`, or a particular RAILS_ENV, or is it simply `bundle exec rspec` causing the issue?

**Comment by zs787** [comment]
Thanks for the details.

As for how I run it all I do is set `COVERAGE=true` then we invoke/require the coverage tools like deep-cover and simplecov. The command would then be `COVERAGE=true bundle exec rspec`.

Running it without requiring deep-cove or without coverage will load the tests files and everything would just run normally.

**Comment by marcandre** [comment]
Update: I'm able to reproduce it! :-) Thanks @zs787 for the info. I'll keep you updated

**Comment by zs787** [comment]
> Update: I'm able to reproduce it! :-) Thanks @zs787 for the info. I'll keep you updated

Thats great and thanks for keeping me updated.

**Comment by marcandre** [comment]
So it ended up being much simpler than I thought. We simply screwed up and made `Kernel::exec` public. That confuses `pg` which also deals with its own `Connection#exec` method.

Fixed in 0.7.3.

Thanks for reporting this bug and your patience in guiding us towards its source.

**Comment by zs787** [comment]
@marcandre No problem and thank you for your help and fast response to get this fixed.
